zap/tests/irgen/mod.rs

```diff
@@ -0,0 +1,217 @@
+		self.push_stmts(statements);
+		self.push_line("return value");
+		self.dedent();
+		self.push_line("end");
+	}
+
+	fn push_event_callback(&mut self, parameters: &Vec<Parameter<'src>>, default_value_index: &str) {
+		self.push_line("do");
+		self.indent();
+
+		let ser_names = parameters
+			.iter()
+			.enumerate()
+			.map(|parameters| format!("ser_{}", parameters.0 + 1))
+			.collect::<Vec<String>>();
+
+		let des_names = parameters
+			.iter()
+			.enumerate()
+			.map(|parameters| format!("des_{}", parameters.0 + 1))
+			.collect::<Vec<String>>();
+
+		let default_values = self
+			.default_values
+			.get(default_value_index)
+			.unwrap_or_else(|| panic!("No default values for {default_value_index}"))
+			.clone();
+
+		for (ser_name, default_value) in ser_names.iter().zip(default_values) {
+			self.push_line(&format!("local {ser_name} = {default_value}"));
+		}
+
+		for des_name in &des_names {
+			self.push_line(&format!("local {des_name}"));
+		}
+
+		let ser_statements = ser::gen(
+			parameters.iter().map(|parameter| &parameter.ty),
+			&ser_names,
+			self.config.write_checks,
+			&mut self.var_occurrences,
+		);
+
+		self.push_stmts(&ser_statements);
+		self.push_line("incoming_buff = outgoing_buff");
+		self.push_line("incoming_inst = outgoing_inst");
+		self.push_line("incoming_read = 0");
+		self.push_line("incoming_ipos = 0");
+		self.push_line("load_empty()");
+
+		let des_statements = des::gen(
+			parameters.iter().map(|parameter| &parameter.ty),
+			&des_names,
+			self.config.write_checks,
+			&mut self.var_occurrences,
+		);
+
+		self.push_stmts(&des_statements);
+
+		for (ser_name, des_name) in ser_names.iter().zip(des_names.iter()) {
+			self.push_line(&format!("assert({ser_name} == {des_name})"));
+		}
+
+		self.dedent();
+		self.push_line("end");
+	}
+
+	pub fn output(mut self) -> String {
+		self.push(include_str!("./polyfill.luau"));
+		self.push(include_str!("../../src/output/luau/base.luau"));
+
+		for tydecl in &self.config.tydecls {
+			self.push_tydecl(tydecl);
+		}
+
+		for evdecl in self.config.evdecls.iter() {
+			self.push_event_callback(&evdecl.data, evdecl.name);
+		}
+
+		for fndecl in self.config.fndecls.iter() {
+			self.push_event_callback(&fndecl.args, &format!("{}__ARGS", fndecl.name));
+
+			if let Some(rets) = &fndecl.rets {
+				let parameters = &rets
+					.iter()
+					.map(|ty| Parameter {
+						name: None,
+						ty: ty.to_owned(),
+					})
+					.collect::<Vec<Parameter<'src>>>();
+
+				self.push_event_callback(parameters, &format!("{}__RETS", fndecl.name));
+			}
+		}
+
+		self.buf
+	}
+}
+
+#[tokio::test]
+async fn test_function() {
+	let (config, reports) = parse(include_str!("../files/function.zap"));
+
+	assert!(config.is_some());
+	assert!(reports.is_empty());
+
+	let default_values: HashMap<&str, Vec<&str>> = HashMap::from([
+		("Test__ARGS", vec!["0", "\"foo\""]),
+		("Test__RETS", vec!["\"Success\"", "\"bar\""]),
+	]);
```

Comment:
Not the happiest with how `default_values` is done in this file. Let me know if you have any better ideas.